server/src/main/java/org/opensearch/cluster/decommission/DecommissionHelper.java

```diff
@@ -0,0 +1,55 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.cluster.decommission;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.opensearch.cluster.ClusterStateTaskConfig;
+import org.opensearch.cluster.ClusterStateTaskListener;
+import org.opensearch.cluster.coordination.NodeRemovalClusterStateTaskExecutor;
+import org.opensearch.cluster.node.DiscoveryNode;
+import org.opensearch.cluster.service.ClusterService;
+import org.opensearch.common.Priority;
+
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+
+public class DecommissionHelper {
+
+    private static final Logger logger = LogManager.getLogger(DecommissionHelper.class);
+
+    private final NodeRemovalClusterStateTaskExecutor nodeRemovalExecutor;
+    private final ClusterService clusterService;
+
+    DecommissionHelper(
+        ClusterService clusterService,
+        NodeRemovalClusterStateTaskExecutor nodeRemovalClusterStateTaskExecutor
+    ) {
+        this.nodeRemovalExecutor = nodeRemovalClusterStateTaskExecutor;
+        this.clusterService = clusterService;
+    }
+
+    public void handleNodesDecommissionRequest(List<DiscoveryNode> nodesToBeDecommissioned, String reason) {
+        final Map<NodeRemovalClusterStateTaskExecutor.Task, ClusterStateTaskListener> nodesDecommissionTasks = new LinkedHashMap<>();
+        nodesToBeDecommissioned.forEach(discoveryNode -> {
+            final NodeRemovalClusterStateTaskExecutor.Task task = new NodeRemovalClusterStateTaskExecutor.Task(
+                discoveryNode, reason
+            );
+            nodesDecommissionTasks.put(task, nodeRemovalExecutor);
```

Comment:
Should we add a ClusterStateTaskListener that does things the way we want to happen. lets say on failure we should basically be changing the decommission status to reflect the state. and similarly on success/response we should be clearing the exclusion.

Reply:
Adding a clusterStateObserver which would be listening to any cluster state update until a timeout and checks if the updated state has any decommissioned nodes or not. On a successful match or a timeout we will mark the status as SUCCESS or FAILURE respectively. Eg. - 
```
       final ClusterStateObserver observer = new ClusterStateObserver(
            clusterService,
            TimeValue.timeValueSeconds(30L),
            logger,
            threadPool.getThreadContext()
        );

        final Predicate<ClusterState> allDecommissionedNodesRemoved = clusterState -> {
            List<DiscoveryNode> nodesWithDecommissionAttribute = nodesWithDecommissionAttribute(clusterState, decommissionAttribute);
            return nodesWithDecommissionAttribute.size() == 0;
        };

        observer.waitForNextChange(new ClusterStateObserver.Listener() {
            @Override
            public void onNewClusterState(ClusterState state) {
                clearVotingConfigAfterSuccessfulDecommission();
                updateMetadataWithDecommissionStatus(DecommissionStatus.DECOMMISSION_SUCCESSFUL);
            }

            @Override
            public void onClusterServiceClose() {
            }

            @Override
            public void onTimeout(TimeValue timeout) {
                clearVotingConfigAfterSuccessfulDecommission();
                updateMetadataWithDecommissionStatus(DecommissionStatus.DECOMMISSION_FAILED);
            }
        }, allDecommissionedNodesRemoved, TimeValue.timeValueSeconds(30L));
```

Reply:
The `DecommissionService` is currently bind to decommissioning an awareness attribute. Whereas the `DecommissionHelper` is more generic in nature and accepts list of nodes that are supposed to be kicked out of the cluster and hence decoupled it.

I don't have a strong opinion though on this and this can be easily abstracted out later if required. Let me know what do you think

Reply:
+1 ```DecommissionService``` shouldnt have binding to awareness attribute.

Reply:
We can keep `DecommissionController` (previously `DecommissionHelper`) more generic in nature and rename `DecommissionService` to `DecommissionAwarenessAttributeService`. LMK your thoughts.